opentracing/src/main/java/opentracing/SpanPropagator.java

```diff
@@ -0,0 +1,53 @@
+/**
+ * Copyright 2016 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package opentracing;
+
+import java.util.Map;
+
+/**
+ * Encodes or Decodes a {@linkplain Span} in binary or text formats.
+ *
+ * <p>The toXXX methods are expected to serialize spans into a pair of values representing
+ * separately the span / span identity, and the trace attributes. This is done specifically
+ * for binary protocols that may represent tracing identity in a dedicated fixed-length slot in the
+ * binary message format, so that it can be inspected efficiently by the middleware / routing layers
+ * without parsing the whole message.
+ */
+public interface SpanPropagator {
```

Comment:
While there's simplifications in this class that can make the API a lot easier to use, I'm not comfortable with the design of SpanPropagator.

The pushing of `byte[]` and `Map<String,String>` seems too presumptuous to me. How do we know if this is what all protocol middleware frameworks will work with? I do believe this needs to be a more SPI interface to the concept of a middleware framework, and not a presumption that we will always be able to reach into its guts and get a hold of the wire's byte or character stream.

I raised this concern before, and implemented stuff like this before. My suggestion is to keep brainstorming on this problem. The Apache Request framework did something similar here so to adapt to different http middleware frameworks: Servlet, Portlet, JSP, Mustache, Velocity and FreeMarker, I'm aware that its tackling a problem in a different domain but it still draws strong enough parallels.

The Apache Request framework does it like with the [AttributeExtractor](http://tiles.apache.org/tiles-request/apidocs/org/apache/tiles/request/attribute/AttributeExtractor.html) class. By letting the unknown be the subclass instead of the parameter the api design gets to focus on what it does know; **each protocol middleware framework that we support must provide a way (a subclass) to extract the span identity and trace attributes out for us.**